traffic_ops/client/capability.go

```diff
@@ -0,0 +1,121 @@
+package client
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * 	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import "encoding/json"
+import "fmt"
+import "net"
+import "net/http"
+
+import "github.com/apache/trafficcontrol/lib/go-tc"
+
+const API_v4_CAPABILITIES = "/api/1.4/capabilities"
+
+// CreateCapability creates the passed capability.
+func (to *Session) CreateCapability(c tc.Capability) (tc.Alerts, ReqInf, error) {
+	var remoteAddr net.Addr
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+
+	reqBody, err := json.Marshal(c)
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+
+	resp, remoteAddr, err := to.request(http.MethodPost, API_v4_CAPABILITIES, reqBody)
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+	defer resp.Body.Close()
+	reqInf.RemoteAddr = remoteAddr
+
+	var alerts tc.Alerts
+	err = json.NewDecoder(resp.Body).Decode(&alerts)
+	return alerts, reqInf, err
+}
+
+// ReplaceCapabilityByName replaces the capability named 'name' with the one passed as 'c'.
+func (to *Session) ReplaceCapabilityByName(name string, c tc.Capability) (tc.Alerts, ReqInf, error) {
+	var remoteAddr net.Addr
+	reqInf := ReqInf{CacheHitStatus: CacheHitStatusMiss, RemoteAddr: remoteAddr}
+
+	reqBody, err := json.Marshal(c)
+	if err != nil {
+		return tc.Alerts{}, reqInf, err
+	}
+
+	endpoint := fmt.Sprintf("%s?name=%s", API_v4_CAPABILITIES, name)
```

Comment:
I know we do not do this across any of our client code so can ignore this commentbut we really should leverage the net/url lib to build up our query parameters so they are url encoded.

Reply:
yes, we really should